crypto/rand/rand_unix.c

```diff
@@ -192,22 +192,30 @@ void rand_pool_keep_random_devices_open(int keep)
 #  if (defined(__FreeBSD__) || defined(__NetBSD__)) && defined(KERN_ARND)
 /*
  * sysctl_random(): Use sysctl() to read a random number from the kernel
- * Returns the size on success, 0 on failure.
+ * Returns the number of bytes returned in buf on success, -1 on failure.
  */
-static size_t sysctl_random(char *buf, size_t buflen)
+static ssize_t sysctl_random(char *buf, size_t buflen)
 {
     int mib[2];
     size_t done = 0;
     size_t len;
 
+    /* Just to be on the safe side. The expected buflen is much smaller. */
+    if (!ossl_assert(buflen <= OSSL_SSIZE_MAX)) {
+        errno = EINVAL;
+        return -1;
+    }
```

Comment:
Well, this one is actually redundant, because sysctl_random is called from syscall_random, which already performed the check. One can even wonder following. Wouldn't check for buflen <= OSSL_SSIZE_MAX/8 in rand_pool_acquire_entropy be more appropriate? Because it's multiplied by 8 when it's passed to rand_pool_add_end and you don't want that to overflow. Again, I'm not saying that it would, ever, but if we take this path, then it would be only appropriate if we take it all the way. On related note double-check cast in equality test in rand_pool_acquire_entropy...

Reply:
> Well, this one is actually redundant, because sysctl_random is called from syscall_random

Indeed, I considered adding only a comment about it, but then I chose the more defensive solution. I'll remove the redundand check and will also consider your other suggestion about rand_pool_acquire_entropy().